### ISSUE TITLE
Wrong PSNR values

### ISSUE THREAD

**Issue by digitalTrilunaire** [OpenVisualCloud/SVT-AV1#387]
I compared two YUVs (the reference file and the reconstructed file) with ffmpeg and one tool of my company. I can not find the values printed by [SVTEncApp](https://github.com/OpenVisualCloud/SVT-AV1/commit/5fd69642f40655d2ec7ac6ffb8cb2c678650e1e7) either with ffmpeg and with our tool (difference of more than a tenth).
Here is my command with a video taken from xiph website ([objectif-1-fast set](https://media.xiph.org/video/derf/objective-1-fast.tar.gz)) :
`SvtAv1EncApp -stat-report 1 -i ../Netflix_Crosswalk_1920x1080p_60Hz_i420.yuv -w 1920 -h 1080 -fps 60 -b Netflix_Crosswalk_1920x1080p_60Hz_i420.ivf -o Netflix_Crosswalk_reconstruct_1920x1080p_60Hz_i420.yuv`
Here is the ffmpeg command used:
`ffmpeg -s 1920x1080 -r 60 -i ../Netflix_Crosswalk_1920x1080p_60Hz_i420.yuv -s 1920x1080 -r 60 -i Netflix_Crosswalk_reconstruct_1920x1080p_60Hz_i420.yuv -lavfi psnr -f null -`
The table below gives you the PSNR with the three tools:

| Tool      | Avg PSNR Y | Avg PSNR U | Avg PSNR V | Overall PSNR Y | Overall PSNR U | Overall PSNR V |
|-----------|------------|------------|------------|----------------|----------------|----------------|
| [SVTEncApp](https://github.com/OpenVisualCloud/SVT-AV1/commit/5fd69642f40655d2ec7ac6ffb8cb2c678650e1e7) | 41.15      | 45.69 | 47.17      | N/A            | N/A            | N/A            |
| FFmpeg    | N/A        | N/A        | N/A        | 40.859242    | 45.536283  | 47.044303  |
| testpsnr  | 41.0125    | 45.5633    | 47.0744    | 40.8592   | 45.5363 | 47.0443  |

**Comment by hassount** [comment]
@digitalTrilunaire 
Thank you for reporting, this is actually an unexpected behavior. I ll look into the differences and get back to you. there's a possibility that some features are causing the original pixels to change which is causing the difference.

**Comment by hassount** [comment]
@digitalTrilunaire 
so the difference is due to the averaging order, svt is averaging the PSNR numbers to get the average per clip and ffmpeg seems to be averaging the sse numbers then calculating the average psnr. 
the per frame psnrs are exactly the same, the clip psnr is different.

**Comment by digitalTrilunaire** [comment]
@hassount , yes there is a difference in the computation of psnr between FFmpeg and SVT-AV1, in our company we difference them with Average PSNR (the most commonly used) and overall PSNR (outped by ffmpeg).
Here are the results from 46df8a8 with the same test (first comment) :

| Tool      | Avg PSNR Y | Avg PSNR U | Avg PSNR V | Overall PSNR Y | Overall PSNR U | Overall PSNR V |
|-----------|------------|------------|------------|----------------|----------------|----------------|
| SVTEncApp 46df8a8 | 41.36     | 45.91| 47.43     | N/A            | N/A            | N/A            |
| FFmpeg    | N/A        | N/A        | N/A        | 41.082146    | 45.780429  | 47.321949  |
| testpsnr  | 41.2294    | 45.8024   | 47.3459  | 41.0821 | 45.7804 | 47.3219 |

Here are the results from adb8939 (the psnr values are still different) :

| Tool      | Avg PSNR Y | Avg PSNR U | Avg PSNR V | Overall PSNR Y | Overall PSNR U | Overall PSNR V |
|-----------|------------|------------|------------|----------------|----------------|----------------|
| SVTEncApp 46df8a8 | 41.14     | 45.86 | 47.38     | N/A            | N/A            | N/A            |
| FFmpeg    | N/A        | N/A        | N/A        | 41.082146    | 45.780429  | 47.321949  |
| testpsnr  | 41.2294    | 45.8024   | 47.3459  | 41.0821 | 45.7804 | 47.3219 |

I made differents tests to compute BD-rates and I notice that the psnr values outputed by coding tools are the average PSNR (I just test HM encoder, X265 and AOM), but it's up to the software engineer to compute what he wants :) . You can verify it e.g., with an aom encoding (if you have time), like this one (with aom normative commit 34de074):
`aomenc --tune=psnr --psnr -w 1920 -h 1080 --fps=60/1 --verbose --cpu-used=8 --cdf-update-mode=0 --enable-cdef=0 --end-usage=cq --cq-level=63 --tile-columns=3 -t 4 -o encoded/Netflix_Crosswalk_1920x1080p_60Hz_i420_aom.ivf Netflix_Crosswalk_1920x1080p_60Hz_i420.yuv`
I decoded the reconstructed yuv with dav1d and here are the psnr results

| Tool      | Avg PSNR Y | Avg PSNR U | Avg PSNR V | Overall PSNR Y | Overall PSNR U | Overall PSNR V |
|-----------|------------|------------|------------|----------------|----------------|----------------|
| aom 34de074 | 34.868 | 41.358 | 42.726 | N/A            | N/A            | N/A            |
| FFmpeg    | N/A        | N/A        | N/A        | 34.817416 | 41.521223  | 42.747534 |
| testpsnr  | 34.8869    | 41.5339  | 42.7546 | 34.8174 | 41.5212 | 42.7475 |

You can notice that the computed value is the Average PSNR. For the Avg Y value there are a difference of only two hundredth (negligeable when you compute BD rates). I'll try with the aom v1.0 and update the results.

**Comment by digitalTrilunaire** [comment]
As promise here are the results with [aom errata](https://aomedia.googlesource.com/aom/+/refs/tags/v1.0.0-errata1) tag. I modify my command and activate multithreading of aom:
`aomenc --tune=psnr --psnr -w 1920 -h 1080 --fps=60/1 --verbose --cpu-used=8 --end-usage=cq --cq-level=63 --tile-columns=3 -t 4 -o encoded/Netflix_Crosswalk_1920x1080p_60Hz_i420_aom.ivf Netflix_Crosswalk_1920x1080p_60Hz_i420.yuv`

I decoded it with aomdec rather than dav1d:
`aomdec --i420 Netflix_Crosswalk_1920x1080p_60Hz_i420_aom.ivf -o Netflix_Crosswalk_aomErrata_reco_1920x1080p_60Hz_i420.yuv`

The results are more reliables than the previous table

| Tool      | Avg PSNR Y | Avg PSNR U | Avg PSNR V | Overall PSNR Y | Overall PSNR U | Overall PSNR V |
|-----------|------------|------------|------------|----------------|----------------|----------------|
| aom add4b15 | 35.287 | 41.762 | 43.128 | N/A            | N/A            | N/A            |
| FFmpeg    | N/A        | N/A        | N/A        | 35.223099 | 41.752415 | 43.123057 |
| testpsnr  | 35.2867  | 41.7616 | 43.1277 | 35.2231 | 41.7524 | 43.1231 |

You can notice that the coder output the average value rather than overall value.

**Comment by joelsole** [comment]
> @digitalTrilunaire
> so the difference is due to the averaging order, svt is averaging the PSNR numbers to get the average per clip and ffmpeg seems to be averaging the sse numbers then calculating the average psnr.
> the per frame psnrs are exactly the same, the clip psnr is different.

Both methods (Classic PSNR vs True PSNR) are being used, not always explaining which one, which creates a bit of confusion.

Externally of SVT-AV1, I was computing PSNR both ways. Both numbers tend to be similar unless there is large variance on the per-frame SSE. So having both might help identify variance in the frame-qualities. 

I would argue to have both reported by SVT-AV1. 

More info from a paper last year by Netflix:

One can easily show that Classic PSNR can also be calculated by taking the geometric mean of MSE values for the Y component first, and then expressing that mean in dB. As a result, and focusing on the Y-component for a moment, one can prove (inequalities among harmonic, geometric and arithmetic means of sequences) that the Classic-PSNR value is always higher from the corresponding True-PSNR value. In practice, since Cb- and Cr-PSNR values are typically higher than Y-PSNR for a given frame, there is no mathematically guarantee on such inequality. Moreover, we can show that large differences in MSE values among different frames in a sequence result in a higher discrepancy between True PSNR and Classic PSNR values. Given that there is a number of publications that prefers one over the other, we decided to calculate and use both these metrics for encoded shots. We hope that, through subjective validation, the video compression community can settle on a single PSNR metric that can be used consistently for reporting research results.

**Comment by digitalTrilunaire** [comment]
@joelsole interesting comment, I agree with your argue to have both reported value. For a person like me I use what you're calling classical PSNR, but other people may want to use the True PSNR. 

> More info from a paper last year by Netflix:

Could you provide us the link of this paper, I'm interested about it :)

**Comment by joelsole** [comment]
Here: http://spie.org/Publications/Proceedings/Paper/10.1117/12.2322118?SSO=1
It seems it is not free, though
